### ISSUE TITLE
Silk.NET.Input SdlKeyboard.Keys not all mapped

### ISSUE THREAD

**Issue by usertoroot** [dotnet/Silk.NET#846]
I am using Silk.NET for an open source project called NetInput (not live yet) which does what software like Parsec does (stream input to PC and stream video back). With the only difference that it is input only so without the video (video is sent using another way).

Essentially providing a low latency way of sending input over the network.

I was almost fully able to implement this using Silk.NET.Input, absolutely incredible work there. There's only 1 issue I ran into, some keys weren't working. After downloading Silk.NET and debugging I encountered the following:
https://github.com/dotnet/Silk.NET/blob/bf3a2cbce752e6baa87c2dc41fe451f38cf90f60/src/Input/Silk.NET.Input.Sdl/SdlKeyboard.Keys.cs#L13

In the following code, as specified in the comment, the enum is designed for UK-style QWERTY keyboards. However for us US international users this means that the keys ` ' and numpad . are not working because they do not exist in the key map.

As specified in:
https://github.com/dotnet/Silk.NET/blob/bf3a2cbce752e6baa87c2dc41fe451f38cf90f60/src/Input/Silk.NET.Input.Sdl/SdlKeyboard.cs#L43

I was thinking of contributing a fix, but I did not even know what a generic fix would look like and if such a fix even exists. In windows for example they just uses VirtualKey codes where these keys just map to OEM keys:
```
Key.GraveAccent => WindowsVirtualKeyCode.OEM_3,
Key.Apostrophe => WindowsVirtualKeyCode.OEM_7,
```

At least to me the wrong behavior seems to be that if it doesn't exist in the map, don't throw the event at all because it doesn't even give the user of the library a chance to handle the keyboard event.

Thank you for your hard work. If you give me some feedback perhaps I can make a pull request for a possible fix.

**Comment by usertoroot** [comment]
Perhaps, the easiest fix I can think of (with minimal change) would be that if the key does not exist in the map, send Key.Unknown with the raw key code. This at least gives the user the ability to handle it?

**Comment by Perksey** [comment]
Hey, thanks so much!

It is known that our SDL backend is bugged, and we recommend using GLFW on all desktop platforms. The Silk.NET.Windowing package, in .NET 6, will switch between the two backend packages as it sees fit.

We have an idea to fix the problem in #737 but this is a community-driven issue and no one has picked it up yet. 3.0 will also fix this given it’s scancode-first.

I recommend using the scancode to do your own mapping. The Key enum will always be limited by design.

Hopefully this helps, probably not the answer you were hoping for but if we were to handle regional keyboard layouts we’d be committing to a lot!

**Comment by usertoroot** [comment]
I believe the reason I ended up in SDL was because of a bug with gamepads on GLFW. I will take a look at this at a later time and perhaps see if I can fix it.

The scancode-first approach sounds like a good idea. That way you allow the user to convert it if required using APIs like 
https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-mapvirtualkeya

The way Microsoft and other software I've seen do it is by exposing only the common keys in the enum which is in this case `SdlKeyboard.Keys` other keys will either map to reserved or OEM.

https://docs.microsoft.com/en-us/windows/win32/inputdev/virtual-key-codes

Am I understanding correctly that this is already part of 3.0? (did a quick search but wasn't able to find it)

If this is not already implemented hereby I am proposing a tiny change to SdlKeyboard and GlfwKeyboard to implement:

If it doesn't exist in the map do not return, map to `Keys.Unknown`

in code:

from

```
                    if (@event.Key.Repeat != 1 && _keyMap.TryGetValue((KeyCode) @event.Key.Keysym.Sym, out var key))
                    {
                        _keysDown.Add(key);
                        KeyDown?.Invoke(this, key, (int) @event.Key.Keysym.Scancode);
                    }
```

to

```
                    if (@event.Key.Repeat != 1)
                    {
                        Key keyDown;
                        if (_keyMap.TryGetValue((KeyCode) @event.Key.Keysym.Sym, out var key))
                        {
                            _keysDown.Add(key);
                            keyDown = key;
                        }
                        else
                        {
                            keyDown = Keys.Unknown;
                        }

                        KeyDown?.Invoke(this, keyDown, (int) @event.Key.Keysym.Scancode);
                    }
```

This seems like a minimal approach and implements what you suggest. If you agree I will implement it and PR.

**Comment by Perksey** [comment]
That seems like a good change.

And yes this has already been _designed_ into 3.0 (no implementation work has actually started on 3.0, we've only just got the designs approved in our [community meeting](https://www.youtube.com/watch?v=dac3t0oh3VU))

**Comment by Perksey** [comment]
Fixed in #848, thanks!
